compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,864 @@
+                .cast<MemRefType>();
+      else
+        dst_memref_ty =
+            memref::SubViewOp::inferResultType(
+                dst_memref_ty, subviewOp.getStaticOffsets(),
+                subviewOp.getStaticSizes(), subviewOp.getStaticStrides())
+                .cast<MemRefType>();
+    }
+  }
+
+  if (src_ancestor_memref_ops.size()) {
+    extractStridesFromMemrefType(src_memref_ty, rewriter, src_strides);
+    extractSizesFromMemrefType(src_memref_ty, rewriter, src_sizes);
+  }
+  if (dst_ancestor_memref_ops.size()) {
+    extractStridesFromMemrefType(dst_memref_ty, rewriter, dst_strides);
+    extractSizesFromMemrefType(dst_memref_ty, rewriter, dst_sizes);
+  }
+
+  SmallVector<Value, 4> deps;
+  SmallVector<Type, 4> tys;
+
+  if (failed(canonicalizeAIRDmaOperands(rewriter, src_offsets, src_sizes,
+                                        src_strides,
+                                        src.getType().cast<MemRefType>())) ||
+      failed(canonicalizeAIRDmaOperands(rewriter, dst_offsets, dst_sizes,
+                                        dst_strides,
+                                        dst.getType().cast<MemRefType>()))) {
+    assert(false);
+  }
+  auto new_dma = rewriter.create<xilinx::air::DmaMemcpyNdOp>(
+      loc, tys, deps, dst, dst_offsets, dst_sizes, dst_strides, src,
+      src_offsets, src_sizes, src_strides);
+
+  assert(!new_dma.getSrcMemref().getDefiningOp<memref::TransposeOp>());
+  assert(!new_dma.getDstMemref().getDefiningOp<memref::TransposeOp>());
+
+  dmaOp->erase();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Pass
+//===----------------------------------------------------------------------===//
+
+class AMDAIEDecomposeLinalgExtPackUnPackToAIRPass
+    : public AMDAIEDecomposeLinalgExtPackUnPackToAIRBase<
+          AMDAIEDecomposeLinalgExtPackUnPackToAIRPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, func::FuncDialect,
+                    arith::ArithDialect, scf::SCFDialect, memref::MemRefDialect,
+                    IREE::LinalgExt::IREELinalgExtDialect,
+                    xilinx::air::airDialect>();
+  }
+
+  AMDAIEDecomposeLinalgExtPackUnPackToAIRPass() = default;
+  AMDAIEDecomposeLinalgExtPackUnPackToAIRPass(
+      const AMDAIEDecomposeLinalgExtPackUnPackToAIRPass &pass){};
+
+  void runOnOperation() override;
+};
+
+void AMDAIEDecomposeLinalgExtPackUnPackToAIRPass::runOnOperation() {
+  MLIRContext *ctx = &getContext();
+
+  // Generalization patterns for outer unit dims have higher priority because
+  // they do not generate reshape ops.
+  {
+    RewritePatternSet patterns(ctx);
+    patterns.add<GeneralizePackOpPattern, GeneralizeUnPackOpPattern>(ctx);
+    if (failed(applyPatternsAndFoldGreedily(getOperation(),
+                                            std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
```

Comment:
I haven't looked into details yet. The reason that we have them for tensor version is that we want to get rid of tensor.collpase_shape/expand_shape ops. They could block vectorization and bufferization analysis. Given that you are working on memref's world. Is it okay to just leverage them to LowerPackPattern and LowerUnPackPattern? Perhaps memref.collapse_shape/expand_shape ops are probably okay in your use cases, have you considered that?

Reply:
Thanks for your comment. Yes I did consider memref.collapse_shape but I coundn't get it to work with the test case that I have, as it was complaining about "invalid source layout map or collapsing non-contiguous dims". Separating the `GeneralizePackOpPattern` from `LowerPack` helped us to quickly generate working examples with the application we are looking at.